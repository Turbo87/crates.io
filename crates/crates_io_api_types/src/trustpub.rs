use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, utoipa::ToSchema)]
#[schema(as = GitHubConfig)]
pub struct GitHubConfig {
    #[schema(example = 42)]
    pub id: i32,
    #[schema(example = "regex")]
    #[serde(rename = "crate")]
    pub krate: String,
    #[schema(example = "rust-lang")]
    pub repository_owner: String,
    #[schema(example = 5430905)]
    pub repository_owner_id: i32,
    #[schema(example = "regex")]
    pub repository_name: String,
    #[schema(example = "ci.yml")]
    pub workflow_filename: String,
    #[schema(example = json!(null))]
    pub environment: Option<String>,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Deserialize, utoipa::ToSchema)]
#[schema(as = NewGitHubConfig)]
pub struct NewGitHubConfig {
    #[schema(example = "regex")]
    #[serde(rename = "crate")]
    pub krate: String,
    #[schema(example = "rust-lang")]
    pub repository_owner: String,
    #[schema(example = "regex")]
    pub repository_name: String,
    #[schema(example = "ci.yml")]
    pub workflow_filename: String,
    #[schema(example = json!(null))]
    pub environment: Option<String>,
}
