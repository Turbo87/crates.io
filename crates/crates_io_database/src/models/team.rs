use bon::Builder;
use diesel::prelude::*;
use diesel_async::{AsyncPgConnection, RunQueryDsl};

use crate::models::{Crate, CrateOwner, Owner, OwnerKind};
use crate::schema::{crate_owners, teams};

/// For now, just a GitHub Team. Can be upgraded to other teams
/// later if desirable.
#[derive(Queryable, Identifiable, serde::Serialize, serde::Deserialize, Debug, Selectable)]
pub struct Team {
    /// Unique table id
    pub id: i32,
    /// "github:org:team"
    /// An opaque unique ID, that was at one point parsed out to query GitHub.
    /// We only query membership with github using the github_id, though.
    /// This is the only name we should ever talk to Cargo about.
    pub login: String,
    /// The GitHub API works on team ID numbers. This can change, if a team
    /// is deleted and then recreated with the same name!!!
    pub github_id: i32,
    /// Sugary goodness
    pub name: Option<String>,
    pub avatar: Option<String>,
    /// The GitHub Organization ID this team sits under
    pub org_id: i32,
}

#[derive(Insertable, AsChangeset, Debug, Builder)]
#[diesel(table_name = teams, check_for_backend(diesel::pg::Pg))]
pub struct NewTeam<'a> {
    pub login: &'a str,
    pub github_id: i32,
    pub name: Option<&'a str>,
    pub avatar: Option<&'a str>,
    pub org_id: i32,
}

impl NewTeam<'_> {
    pub async fn create_or_update(&self, conn: &mut AsyncPgConnection) -> QueryResult<Team> {
        use diesel::insert_into;

        insert_into(teams::table)
            .values(self)
            .on_conflict(teams::github_id)
            .do_update()
            .set(self)
            .returning(Team::as_returning())
            .get_result(conn)
            .await
    }
}

impl Team {
    pub async fn owning(krate: &Crate, conn: &mut AsyncPgConnection) -> QueryResult<Vec<Owner>> {
        let base_query = CrateOwner::belonging_to(krate).filter(crate_owners::deleted.eq(false));
        let teams = base_query
            .inner_join(teams::table)
            .select(Team::as_select())
            .filter(crate_owners::owner_kind.eq(OwnerKind::Team))
            .load(conn)
            .await?
            .into_iter()
            .map(Owner::Team);

        Ok(teams.collect())
    }

    /// Splits the login into provider, organization, and team name.
    ///
    /// Returns `None` if the login format is invalid.
    pub fn split_login(&self) -> Option<(&str, &str, &str)> {
        let (provider, rest) = self.login.split_once(':')?;
        let (org, team) = rest.split_once(':')?;
        Some((provider, org, team))
    }

    /// Returns the URL for the team.
    ///
    /// Currently only supports GitHub teams. Returns `None` for other providers
    /// or invalid login formats.
    pub fn url(&self) -> Option<String> {
        let (provider, org, _team) = self.split_login()?;
        match provider {
            "github" => Some(format!("https://github.com/{org}")),
            _ => None,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn new_team(login: &str) -> Team {
        Team {
            id: 1,
            login: login.to_string(),
            github_id: 1000,
            name: None,
            avatar: None,
            org_id: 2000,
        }
    }

    mod split_login {
        use super::*;

        #[test]
        fn valid_login() {
            let team = new_team("github:rust-lang:core");
            assert_eq!(team.split_login(), Some(("github", "rust-lang", "core")));
        }

        #[test]
        fn missing_second_colon() {
            let team = new_team("github:rust-lang");
            assert_eq!(team.split_login(), None);
        }

        #[test]
        fn missing_both_colons() {
            let team = new_team("github");
            assert_eq!(team.split_login(), None);
        }

        #[test]
        fn empty_string() {
            let team = new_team("");
            assert_eq!(team.split_login(), None);
        }

        #[test]
        fn extra_colons() {
            let team = new_team("github:rust-lang:core:extra");
            assert_eq!(
                team.split_login(),
                Some(("github", "rust-lang", "core:extra"))
            );
        }

        #[test]
        fn different_provider() {
            let team = new_team("gitlab:my-org:my-team");
            assert_eq!(team.split_login(), Some(("gitlab", "my-org", "my-team")));
        }
    }

    mod url {
        use super::*;

        #[test]
        fn github_team() {
            let team = new_team("github:rust-lang:core");
            assert_eq!(team.url(), Some("https://github.com/rust-lang".to_string()));
        }

        #[test]
        fn gitlab_team_returns_none() {
            let team = new_team("gitlab:my-org:my-team");
            assert_eq!(team.url(), None);
        }

        #[test]
        fn invalid_format_returns_none() {
            let team = new_team("github:rust-lang");
            assert_eq!(team.url(), None);
        }

        #[test]
        fn empty_login_returns_none() {
            let team = new_team("");
            assert_eq!(team.url(), None);
        }

        #[test]
        fn github_with_hyphenated_org() {
            let team = new_team("github:test-org:core");
            assert_eq!(team.url(), Some("https://github.com/test-org".to_string()));
        }
    }
}
