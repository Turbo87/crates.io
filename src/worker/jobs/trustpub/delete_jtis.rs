use crate::worker::Environment;
use crates_io_database::schema::trustpub_used_jtis;
use crates_io_worker::BackgroundJob;
use diesel::prelude::*;
use diesel_async::RunQueryDsl;
use serde::{Deserialize, Serialize};
use std::sync::Arc;

/// A background job that deletes expired JSON Web Token IDs (JTIs)
/// tokens from the database.
#[derive(Deserialize, Serialize)]
pub struct DeleteExpiredJtis;

impl BackgroundJob for DeleteExpiredJtis {
    const JOB_NAME: &'static str = "trustpub::delete_expired_jtis";
    const DEDUPLICATED: bool = true;

    type Context = Arc<Environment>;

    async fn run(&self, ctx: Self::Context) -> anyhow::Result<()> {
        let mut conn = ctx.deadpool.get().await?;

        diesel::delete(trustpub_used_jtis::table)
            .filter(trustpub_used_jtis::expires_at.lt(diesel::dsl::now))
            .execute(&mut conn)
            .await?;

        Ok(())
    }
}
