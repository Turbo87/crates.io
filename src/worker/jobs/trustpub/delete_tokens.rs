use crate::worker::Environment;
use crates_io_database::schema::trustpub_tokens;
use crates_io_worker::BackgroundJob;
use diesel::prelude::*;
use diesel_async::RunQueryDsl;
use serde::{Deserialize, Serialize};
use std::sync::Arc;

/// A background job that deletes expired temporary access
/// tokens from the database.
#[derive(Deserialize, Serialize)]
pub struct DeleteExpiredTokens;

impl BackgroundJob for DeleteExpiredTokens {
    const JOB_NAME: &'static str = "trustpub::delete_expired_tokens";

    type Context = Arc<Environment>;

    async fn run(&self, ctx: Self::Context) -> anyhow::Result<()> {
        let mut conn = ctx.deadpool.get().await?;

        diesel::delete(trustpub_tokens::table)
            .filter(trustpub_tokens::expires_at.lt(diesel::dsl::now))
            .execute(&mut conn)
            .await?;

        Ok(())
    }
}
