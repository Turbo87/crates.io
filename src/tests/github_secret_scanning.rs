use crate::builders::CrateBuilder;
use crate::util::MockRequestExt;
use crate::util::insta::api_token_redaction;
use crate::{RequestHelper, TestApp};
use base64::{Engine as _, engine::general_purpose};
use chrono::{TimeDelta, Utc};
use claims::assert_ok;
use crates_io::util::token::HashedToken;
use crates_io::{models::ApiToken, schema::api_tokens};
use crates_io_database::models::CrateOwner;
use crates_io_database::models::trustpub::NewToken;
use crates_io_database::schema::trustpub_tokens;
use crates_io_github::{GitHubPublicKey, MockGitHubClient};
use crates_io_trustpub::access_token::AccessToken;
use diesel::prelude::*;
use diesel_async::RunQueryDsl;
use googletest::prelude::*;
use insta::{assert_json_snapshot, assert_snapshot};
use p256::ecdsa::{Signature, SigningKey, signature::Signer};
use p256::pkcs8::DecodePrivateKey;
use secrecy::ExposeSecret;
use std::sync::LazyLock;

static URL: &str = "/api/github/secret-scanning/verify";

// Test request payload for GitHub secret scanning
static GITHUB_ALERT: &[u8] =
    br#"[{"token":"some_token","type":"some_type","url":"some_url","source":"some_source"}]"#;

/// Generate a GitHub alert with a given token
fn github_alert_with_token(token: &str) -> Vec<u8> {
    format!(
        r#"[{{"token":"{token}","type":"some_type","url":"some_url","source":"some_source"}}]"#,
    )
    .into_bytes()
}

/// Private key for signing payloads (ECDSA P-256)
///
/// Generated specifically for testing - do not use in production.
///
/// This corresponds to the public key below and is used to generate valid signatures
static PRIVATE_KEY: &str = r#"-----BEGIN PRIVATE KEY-----
MIGHAgEAMBMGByqGSM49AgEGCCqGSM49AwEHBG0wawIBAQQgV64BdEFXg9aT/m4p
wOQ/o9WUHxZ6qfBaP3D7Km1TOWuhRANCAARYKkbkTbIr//8klg1CMYGQIwtlfNd4
JQYV5+q0s3+JnBSLb1/sx/lEDzmMVZQIZQrACUHFW4UVdmox2NvmNWyy
-----END PRIVATE KEY-----"#;

/// Public key (corresponds to the private key above)
static PUBLIC_KEY: &str = r#"-----BEGIN PUBLIC KEY-----
MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEWCpG5E2yK///JJYNQjGBkCMLZXzX
eCUGFefqtLN/iZwUi29f7Mf5RA85jFWUCGUKwAlBxVuFFXZqMdjb5jVssg==
-----END PUBLIC KEY-----"#;

/// Public key identifier (SHA256 hash of the DER-encoded public key)
static KEY_IDENTIFIER: &str = "2aafbbe2d329af78d875cd2dd0291048799176466844315b6a846d6e12aa26ca";

/// Signing key derived from the private key
static SIGNING_KEY: LazyLock<SigningKey> =
    LazyLock::new(|| SigningKey::from_pkcs8_pem(PRIVATE_KEY).unwrap());

/// Generate a signature for the payload using our private key
fn sign_payload(payload: &[u8]) -> String {
    let signature: Signature = SIGNING_KEY.sign(payload);
    general_purpose::STANDARD.encode(signature.to_der())
}

/// Generate a new Trusted Publishing token and its SHA256 hash
fn generate_trustpub_token() -> (String, Vec<u8>) {
    let token = AccessToken::generate();
    let finalized_token = token.finalize().expose_secret().to_string();
    let hashed_token = token.sha256().to_vec();
    (finalized_token, hashed_token)
}

/// Create a new Trusted Publishing token in the database
async fn insert_trustpub_token(
    conn: &mut diesel_async::AsyncPgConnection,
    crate_ids: &[i32],
) -> QueryResult<String> {
    let (token, hashed_token) = generate_trustpub_token();

    let new_token = NewToken {
        expires_at: Utc::now() + TimeDelta::minutes(30),
        hashed_token: &hashed_token,
        crate_ids,
        trustpub_data: None,
    };

    new_token.insert(conn).await?;

    Ok(token)
}

fn github_mock() -> MockGitHubClient {
    let mut mock = MockGitHubClient::new();

    mock.expect_public_keys().returning(|_, _| {
        let key = GitHubPublicKey {
            key_identifier: KEY_IDENTIFIER.to_string(),
            key: PUBLIC_KEY.to_string(),
            is_current: true,
        };

        Ok(vec![key])
    });

    mock
}

#[tokio::test(flavor = "multi_thread")]
async fn github_secret_alert_revokes_token() {
    let (app, anon, user, token) = TestApp::init()
        .with_github(github_mock())
        .with_token()
        .await;
    let mut conn = app.db_conn().await;

    // Ensure no emails were sent up to this point
    assert_eq!(app.emails().await.len(), 0);

    // Ensure that the token currently exists in the database
    let tokens: Vec<ApiToken> = assert_ok!(
        ApiToken::belonging_to(user.as_model())
            .select(ApiToken::as_select())
            .filter(api_tokens::revoked.eq(false))
            .load(&mut conn)
            .await
    );
    assert_that!(tokens, len(eq(1)));
    assert_eq!(tokens[0].name, token.as_model().name);

    // Set token to expected value in signed request
    let hashed_token = HashedToken::hash("some_token");
    diesel::update(api_tokens::table)
        .set(api_tokens::token.eq(hashed_token))
        .execute(&mut conn)
        .await
        .unwrap();

    let mut request = anon.post_request(URL);
    *request.body_mut() = GITHUB_ALERT.into();
    request.header("GITHUB-PUBLIC-KEY-IDENTIFIER", KEY_IDENTIFIER);
    request.header("GITHUB-PUBLIC-KEY-SIGNATURE", &sign_payload(GITHUB_ALERT));
    let response = anon.run::<()>(request).await;
    assert_snapshot!(response.status(), @"200 OK");
    assert_json_snapshot!(response.json());

    // Ensure that the token was revoked
    let tokens: Vec<ApiToken> = assert_ok!(
        ApiToken::belonging_to(user.as_model())
            .select(ApiToken::as_select())
            .filter(api_tokens::revoked.eq(false))
            .load(&mut conn)
            .await
    );
    assert_that!(tokens, is_empty());

    let tokens: Vec<ApiToken> = assert_ok!(
        ApiToken::belonging_to(user.as_model())
            .select(ApiToken::as_select())
            .filter(api_tokens::revoked.eq(true))
            .load(&mut conn)
            .await
    );
    assert_that!(tokens, len(eq(1)));

    // Ensure exactly one email was sent
    assert_snapshot!(app.emails_snapshot().await);
}

#[tokio::test(flavor = "multi_thread")]
async fn github_secret_alert_for_revoked_token() {
    let (app, anon, user, token) = TestApp::init()
        .with_github(github_mock())
        .with_token()
        .await;
    let mut conn = app.db_conn().await;

    // Ensure no emails were sent up to this point
    assert_eq!(app.emails().await.len(), 0);

    // Ensure that the token currently exists in the database
    let tokens: Vec<ApiToken> = assert_ok!(
        ApiToken::belonging_to(user.as_model())
            .select(ApiToken::as_select())
            .filter(api_tokens::revoked.eq(false))
            .load(&mut conn)
            .await
    );
    assert_that!(tokens, len(eq(1)));
    assert_eq!(tokens[0].name, token.as_model().name);

    // Set token to expected value in signed request and revoke it
    let hashed_token = HashedToken::hash("some_token");
    diesel::update(api_tokens::table)
        .set((
            api_tokens::token.eq(hashed_token),
            api_tokens::revoked.eq(true),
        ))
        .execute(&mut conn)
        .await
        .unwrap();

    let mut request = anon.post_request(URL);
    *request.body_mut() = GITHUB_ALERT.into();
    request.header("GITHUB-PUBLIC-KEY-IDENTIFIER", KEY_IDENTIFIER);
    request.header("GITHUB-PUBLIC-KEY-SIGNATURE", &sign_payload(GITHUB_ALERT));
    let response = anon.run::<()>(request).await;
    assert_snapshot!(response.status(), @"200 OK");
    assert_json_snapshot!(response.json());

    // Ensure that the token is still revoked
    let tokens: Vec<ApiToken> = assert_ok!(
        ApiToken::belonging_to(user.as_model())
            .select(ApiToken::as_select())
            .filter(api_tokens::revoked.eq(false))
            .load(&mut conn)
            .await
    );
    assert_that!(tokens, is_empty());

    let tokens: Vec<ApiToken> = assert_ok!(
        ApiToken::belonging_to(user.as_model())
            .select(ApiToken::as_select())
            .filter(api_tokens::revoked.eq(true))
            .load(&mut conn)
            .await
    );
    assert_that!(tokens, len(eq(1)));

    // Ensure still no emails were sent
    assert_eq!(app.emails().await.len(), 0);
}

#[tokio::test(flavor = "multi_thread")]
async fn github_secret_alert_for_unknown_token() {
    let (app, anon, user, token) = TestApp::init()
        .with_github(github_mock())
        .with_token()
        .await;
    let mut conn = app.db_conn().await;

    // Ensure no emails were sent up to this point
    assert_eq!(app.emails().await.len(), 0);

    // Ensure that the token currently exists in the database
    let tokens: Vec<ApiToken> = assert_ok!(
        ApiToken::belonging_to(user.as_model())
            .select(ApiToken::as_select())
            .filter(api_tokens::revoked.eq(false))
            .load(&mut conn)
            .await
    );
    assert_that!(tokens, len(eq(1)));
    assert_eq!(tokens[0].name, token.as_model().name);

    let mut request = anon.post_request(URL);
    *request.body_mut() = GITHUB_ALERT.into();
    request.header("GITHUB-PUBLIC-KEY-IDENTIFIER", KEY_IDENTIFIER);
    request.header("GITHUB-PUBLIC-KEY-SIGNATURE", &sign_payload(GITHUB_ALERT));
    let response = anon.run::<()>(request).await;
    assert_snapshot!(response.status(), @"200 OK");
    assert_json_snapshot!(response.json());

    // Ensure that the token was not revoked
    let tokens: Vec<ApiToken> = assert_ok!(
        ApiToken::belonging_to(user.as_model())
            .select(ApiToken::as_select())
            .filter(api_tokens::revoked.eq(false))
            .load(&mut conn)
            .await
    );
    assert_that!(tokens, len(eq(1)));
    assert_eq!(tokens[0].name, token.as_model().name);

    // Ensure still no emails were sent
    assert_eq!(app.emails().await.len(), 0);
}

#[tokio::test(flavor = "multi_thread")]
async fn github_secret_alert_invalid_signature_fails() {
    let (_, anon) = TestApp::init().with_github(github_mock()).empty().await;

    // No headers or request body
    let request = anon.post_request(URL);
    let response = anon.run::<()>(request).await;
    assert_snapshot!(response.status(), @"400 Bad Request");

    // Request body but no headers
    let mut request = anon.post_request(URL);
    *request.body_mut() = GITHUB_ALERT.into();
    let response = anon.run::<()>(request).await;
    assert_snapshot!(response.status(), @"400 Bad Request");

    // Headers but no request body
    let mut request = anon.post_request(URL);
    request.header("GITHUB-PUBLIC-KEY-IDENTIFIER", KEY_IDENTIFIER);
    request.header("GITHUB-PUBLIC-KEY-SIGNATURE", &sign_payload(GITHUB_ALERT));
    let response = anon.run::<()>(request).await;
    assert_snapshot!(response.status(), @"400 Bad Request");

    // Request body but only key identifier header
    let mut request = anon.post_request(URL);
    *request.body_mut() = GITHUB_ALERT.into();
    request.header("GITHUB-PUBLIC-KEY-IDENTIFIER", KEY_IDENTIFIER);
    let response = anon.run::<()>(request).await;
    assert_snapshot!(response.status(), @"400 Bad Request");

    // Invalid signature
    let mut request = anon.post_request(URL);
    *request.body_mut() = GITHUB_ALERT.into();
    request.header("GITHUB-PUBLIC-KEY-IDENTIFIER", KEY_IDENTIFIER);
    request.header("GITHUB-PUBLIC-KEY-SIGNATURE", "bad signature");
    let response = anon.run::<()>(request).await;
    assert_snapshot!(response.status(), @"400 Bad Request");

    // Invalid signature that is valid base64
    let mut request = anon.post_request(URL);
    *request.body_mut() = GITHUB_ALERT.into();
    request.header("GITHUB-PUBLIC-KEY-IDENTIFIER", KEY_IDENTIFIER);
    request.header("GITHUB-PUBLIC-KEY-SIGNATURE", "YmFkIHNpZ25hdHVyZQ==");
    let response = anon.run::<()>(request).await;
    assert_snapshot!(response.status(), @"400 Bad Request");
}

#[tokio::test(flavor = "multi_thread")]
async fn github_secret_alert_revokes_trustpub_token() {
    let (app, anon, cookie) = TestApp::init().with_github(github_mock()).with_user().await;
    let mut conn = app.db_conn().await;

    let krate = CrateBuilder::new("foo", cookie.as_model().id)
        .build(&mut conn)
        .await
        .unwrap();

    // Generate a valid Trusted Publishing token
    let token = insert_trustpub_token(&mut conn, &[krate.id]).await.unwrap();

    // Verify the token exists in the database
    let count = trustpub_tokens::table
        .count()
        .get_result::<i64>(&mut conn)
        .await
        .unwrap();
    assert_eq!(count, 1);

    // Send the GitHub alert to the API endpoint
    let mut request = anon.post_request(URL);
    let vec = github_alert_with_token(&token);
    request.header("GITHUB-PUBLIC-KEY-IDENTIFIER", KEY_IDENTIFIER);
    request.header("GITHUB-PUBLIC-KEY-SIGNATURE", &sign_payload(&vec));
    *request.body_mut() = vec.into();
    let response = anon.run::<()>(request).await;
    assert_snapshot!(response.status(), @"200 OK");
    assert_json_snapshot!(response.json(), {
        "[].token_raw" => api_token_redaction()
    });

    // Verify the token was deleted from the database
    let count = trustpub_tokens::table
        .count()
        .get_result::<i64>(&mut conn)
        .await
        .unwrap();
    assert_eq!(count, 0);

    // Ensure an email was sent notifying about the token revocation
    assert_snapshot!(app.emails_snapshot().await);
}

#[tokio::test(flavor = "multi_thread")]
async fn github_secret_alert_for_unknown_trustpub_token() {
    let (app, anon) = TestApp::init().with_github(github_mock()).empty().await;
    let mut conn = app.db_conn().await;

    // Generate a valid Trusted Publishing token but don't insert it into the database
    let (token, _) = generate_trustpub_token();

    // Verify no tokens exist in the database
    let count = trustpub_tokens::table
        .count()
        .get_result::<i64>(&mut conn)
        .await
        .unwrap();
    assert_eq!(count, 0);

    // Send the GitHub alert to the API endpoint
    let mut request = anon.post_request(URL);
    let vec = github_alert_with_token(&token);
    request.header("GITHUB-PUBLIC-KEY-IDENTIFIER", KEY_IDENTIFIER);
    request.header("GITHUB-PUBLIC-KEY-SIGNATURE", &sign_payload(&vec));
    *request.body_mut() = vec.into();
    let response = anon.run::<()>(request).await;
    assert_snapshot!(response.status(), @"200 OK");
    assert_json_snapshot!(response.json(), {
        "[].token_raw" => api_token_redaction()
    });

    // Verify still no tokens exist in the database
    let count = trustpub_tokens::table
        .count()
        .get_result::<i64>(&mut conn)
        .await
        .unwrap();
    assert_eq!(count, 0);

    // Ensure no emails were sent
    assert_eq!(app.emails().await.len(), 0);
}

#[tokio::test(flavor = "multi_thread")]
async fn github_secret_alert_revokes_trustpub_token_multiple_users() {
    let (app, anon) = TestApp::init().with_github(github_mock()).empty().await;
    let mut conn = app.db_conn().await;

    // Create two users
    let user1 = app.db_new_user("user1").await;
    let user2 = app.db_new_user("user2").await;

    // Create two crates
    // User 1 owns both crates 1 and 2
    let crate1 = CrateBuilder::new("crate1", user1.as_model().id)
        .build(&mut conn)
        .await
        .unwrap();
    let crate2 = CrateBuilder::new("crate2", user1.as_model().id)
        .build(&mut conn)
        .await
        .unwrap();

    // Add user 2 as owner of crate2
    CrateOwner::builder()
        .crate_id(crate2.id)
        .user_id(user2.as_model().id)
        .created_by(user1.as_model().id)
        .build()
        .insert(&mut conn)
        .await
        .unwrap();

    // Generate a trusted publishing token that has access to both crates
    let token = insert_trustpub_token(&mut conn, &[crate1.id, crate2.id])
        .await
        .unwrap();

    // Send the GitHub alert to the API endpoint
    let mut request = anon.post_request(URL);
    let vec = github_alert_with_token(&token);
    request.header("GITHUB-PUBLIC-KEY-IDENTIFIER", KEY_IDENTIFIER);
    request.header("GITHUB-PUBLIC-KEY-SIGNATURE", &sign_payload(&vec));
    *request.body_mut() = vec.into();
    let response = anon.run::<()>(request).await;
    assert_snapshot!(response.status(), @"200 OK");
    assert_json_snapshot!(response.json(), {
        "[].token_raw" => api_token_redaction()
    });

    // Take a snapshot of all emails for detailed verification
    assert_snapshot!(app.emails_snapshot().await);
}
