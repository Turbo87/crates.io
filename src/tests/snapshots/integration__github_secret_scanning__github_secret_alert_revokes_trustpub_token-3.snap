---
source: src/tests/github_secret_scanning.rs
expression: app.emails_snapshot().await
---
To: foo@example.com
From: crates.io <noreply@crates.io>
Subject: crates.io: Your Trusted Publishing token has been revoked
Content-Type: text/plain; charset=utf-8
Content-Transfer-Encoding: quoted-printable


GitHub has notified us that one of your crates.io Trusted Publishing tokens has been exposed publicly. We have revoked this token as a precaution.

This token was only authorized to publish the "foo" crate.

Please review your account at https://crates.io and your GitHub repository settings to confirm that no unexpected changes have been made to your crates or trusted publishing configurations.

Source type: some_source

URL where the token was found: some_url

Trusted Publishing tokens are temporary and used for automated publishing from GitHub Actions. If this exposure was unexpected, please review your repository's workflow files and secrets.

--
The crates.io Team
